ls1intum/Artemis
# `Development`: Reimplement course icon file upload

### Checklist
#### General


#### Server
#### Client

### Motivation and Context


Follows #5427 to reimplement all file uploads in our system in a uniform and secure way.

### Description

- Moved `PUT /api/course` to `PUT /api/course/:id` to reflect our standard. I, therefore, also removed the implementation to create a course with PUT. This is not used in our system and only complicates the implementation.
- Modified the create and update endpoints of the course to receive an optional file and save it as the course icon
- Removed the manual upload of the icon and replaced it with a streamlined upload when saving the course.
- Adapted the implementation of detecting changes in the user groups
- I implemented a workaround in the cypress tests because the response's body is an ArrayBuffer instead of an object. Such an issue is already documented (e.g. https://github.com/cypress-io/cypress/issues/17055 for intercept()). Therefore, I implemented a fix that will fail once this gets fixed by Cypress. We can then remove the workaround.

### Steps for Testing

Prerequisites:
- 1 Admin
- 1 Instructor

1. Create Course A with an Admin with a custom image
2. Create Course B with an Admin without a custom image
3. Update a course with an admin adapting the image (expect image updated)
4. Update a course with an admin adapting the user groups (expect groups updated)
5. Update a course with an instructor adapting the image (expect image updated)
6. Update a course with an instructor adapting the image and the user group (you may have to use the API or enable the fields in the DOM. Expect error, no image update)

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|
| CourseResource.java | 72% | 95.25% |
| course-management.service.ts | 77.06% | 90.21% |
| course-update.component.ts | 70.24% | 86.2% |
| file-uploader.service.ts | 41.66% | 53.84% |
| image-cropper.component.ts | 33.5% | 45.25% |

### Screenshots


<details>
<summary>Before</summary>

![](https://gyazo.com/26c3b57220497b5c43f533882b8b156c/raw)
</details>
<details>
<summary>After</summary>

![](https://gyazo.com/42f6228981d1052be602d5cdd72d50e9/raw)
</details>
